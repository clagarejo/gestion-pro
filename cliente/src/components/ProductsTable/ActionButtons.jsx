import { useEffect } from 'react';
import { FaCloudUploadAlt, FaPlus, FaTrash } from 'react-icons/fa';

export const ActionButtons = ({ handleFileUpload, handleOpenModal, handleMassiveDeleteProducts, fileInputRef, selected }) => {
    useEffect(() => {
        const interval = setInterval(() => {
            if (fileInputRef.current) {
                fileInputRef.current.value = "";
            }
        }, 5000);

        return () => clearInterval(interval);
    }, [fileInputRef]);

    return (
        <div className="action-buttons">
            <button onClick={() => fileInputRef.current.click()} className="massive-upload">
                <FaCloudUploadAlt /> Cargar productos
            </button>
            <input
                type="file"
                ref={fileInputRef}
                style={{ display: 'none' }}
                onChange={handleFileUpload}
            />

            <button onClick={() => handleOpenModal()} className="add_product">
                <FaPlus style={{ marginRight: '10px' }} /> Crear producto
            </button>

            {selected.length >= 2 && (
                <button onClick={handleMassiveDeleteProducts} className="masive-delete">
                    <FaTrash style={{ marginRight: '10px' }} /> Eliminar productos
                </button>
            )}
        </div>
    );
};
