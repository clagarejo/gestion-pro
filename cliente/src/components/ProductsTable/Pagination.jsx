import React from 'react';
import PropTypes from 'prop-types';

export const Pagination = ({ totalPages, currentPage, onPageChange }) => {
    return (
        <div className="pagination">
            {Array.from({ length: totalPages }, (_, index) => (
                <button
                    key={index}
                    className={`pagination-button ${currentPage === index + 1 ? 'active' : ''}`}
                    onClick={() => onPageChange(index + 1)}
                >
                    {index + 1}
                </button>
            ))}
        </div>
    );
};

Pagination.propTypes = {
    totalPages: PropTypes.number.isRequired,
    currentPage: PropTypes.number.isRequired,
    onPageChange: PropTypes.func.isRequired
};
