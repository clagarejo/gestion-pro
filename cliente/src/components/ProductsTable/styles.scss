@use '../../styles/variables';

.container {
    padding: 20px;
    max-width: 1400px;
    margin: 0 auto;
    border-radius: 8px;

    .container_bottom {
        display: flex;
        justify-content: space-between;
        margin: 30px 0;

        .masive_delete {
            background-color: variables.$delete-bottom;
            margin-left: 10px;


            &:hover {
                background-color: variables.$delete-bottom-hover;
            }
        }

        .search-products {
            width: 100%;
            max-width: 400px;
            padding: 10px 15px;
            border: 2px solid #ccc;
            border-radius: 5px;
            font-size: 16px;
            color: #333;
            outline: none;
            transition: all 0.3s ease;

            &::placeholder {
                color: #aaa;
            }

            &:focus {
                border-color: #007BFF;
                box-shadow: 0 0 5px rgba(0, 123, 255, 0.5);
            }

            &:hover {
                border-color: #888;
            }

            @media (max-width: 768px) {
                font-size: 14px;
                max-width: 100%;
            }
        }

        button {
            padding: 15px 25px;
            background-color: variables.$primary-color;
            color: #fff;
            border: none;
            border-radius: 5px;
            font-size: 16px;
            cursor: pointer;
            transition: background-color 0.3s ease-in-out;

            &:hover {
                background-color: variables.$primary-color-hover;
            }
        }
    }

    .table {
        width: 100%;
        border-radius: 8px;
        margin-top: 20px;
        cursor: pointer;

        input[type="checkbox"] {
            width: 1.2rem;
            height: 1.2rem;
            cursor: pointer;
        }

        th,
        td {
            padding: 12px 15px;
            text-align: left;
            border-bottom: 1px solid #ddd;
        }

        th {
            font-weight: bold;
        }

        tr:hover {
            background-color: #aeacac;
        }
    }

    .pagination {
        display: flex;
        justify-content: center;
        margin-top: 20px;

        .pagination-button {
            background: #fff;
            border: 1px solid #ddd;
            padding: 8px 12px;
            margin: 0 5px;
            border-radius: 5px;
            cursor: pointer;
            transition: all 0.3s;

            &:hover {
                background: #c4f2ff;
            }

            &.active {
                color: #333;
                border-color: variables.$primary-color;
            }
        }
    }

    @media (max-width: 768px) {
        .table {
            font-size: 12px;

            th,
            td {
                padding: 10px;
            }

            thead {
                display: none;
            }

            tr {
                display: block;
                margin-bottom: 15px;
            }

            td {
                display: block;
                text-align: right;
                padding-left: 50%;
                position: relative;

                &:before {
                    content: attr(data-label);
                    position: absolute;
                    left: 10px;
                    font-weight: bold;
                }
            }
        }

        .pagination {
            flex-wrap: wrap;

            .pagination-button {
                margin-bottom: 10px;
            }
        }
    }
}