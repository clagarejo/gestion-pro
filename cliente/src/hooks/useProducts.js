import { useState, useEffect } from 'react';
import { useProductStore } from '@/store/useProductStore';

export const useProducts = () => {
    const [loading, setLoading] = useState(false);
    const {
        products,
        fetchProducts,
        searchProductsByName,
        selected,
        deleteSelectedProducts,
        processJsonFile,
        toggleProductSelection,
    } = useProductStore();

    useEffect(() => {
        const fetchData = async () => {
            setLoading(true);
            await fetchProducts();
            setLoading(false);
        };

        fetchData();
    }, [fetchProducts]);

    const searchProducts = (searchTerm) => {
        if (searchTerm) {
            searchProductsByName(searchTerm);
        } else {
            fetchProducts();
        }
    };

    const handleSelectAllChange = () => {
        if (selected.length === products.length) {
            toggleProductSelection([]);
        } else {
            const allProductIds = products.map((product) => product._id);
            allProductIds.forEach((id) => toggleProductSelection(id));
        }
    };

    const handleMassiveDeleteProducts = () => {
        setLoading(true)
        deleteSelectedProducts(selected);
        setLoading(false)

    };

    const handleFileUpload = (event) => {
        const file = event.target.files[0];
        if (file) {
            processJsonFile(file);
        }
    };

    return {
        products,
        loading,
        selected,
        searchProducts,
        handleSelectAllChange,
        handleMassiveDeleteProducts,
        handleFileUpload,
    };
};
