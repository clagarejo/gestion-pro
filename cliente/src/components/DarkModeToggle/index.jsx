import React, { useState, useEffect } from "react";
import "./styles.scss";

export const DarkModeToggle = () => {
    const [isDarkMode, setIsDarkMode] = useState(false);

    useEffect(() => {
        const savedMode = localStorage.getItem("theme") || "light";
        setIsDarkMode(savedMode === "dark");
        document.body.classList.toggle("dark-mode", savedMode === "dark");
    }, []);

    const toggleTheme = () => {
        const newTheme = isDarkMode ? "light" : "dark";
        setIsDarkMode(!isDarkMode);
        document.body.classList.toggle("dark-mode", newTheme === "dark");
        localStorage.setItem("theme", newTheme);
    };

    return (
        <button className="theme-toggle" onClick={toggleTheme}>
            {isDarkMode ? "🌙 " : "☀️ "}
        </button>
    );
};

