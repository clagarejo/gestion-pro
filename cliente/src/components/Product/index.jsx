import { FaEdit, FaTrash } from 'react-icons/fa';
import { useProductStore } from '@/store/useProductStore';
import './styles.scss';

export const Product = ({ product, handleOpenModal }) => {
    const productId = product._id;
    const { name, category, price, stock } = product;
    const { deleteProduct } = useProductStore();

    const handleDelete = () => {
        deleteProduct(productId);
    };

    const formatPrice = (price) => {
        return new Intl.NumberFormat('es-CO', {
            style: 'currency',
            currency: 'COP',
            minimumFractionDigits: 0, // Evitar decimales si el precio es entero
        }).format(price);
    };

    return (
        <tr>
            <td>{name}</td>
            <td>{category}</td>
            <td>{formatPrice(price)}(cop)</td>
            <td style={{ textAlign: 'center' }}>{stock}</td>
            <td className="tr-options">
                <button
                    className="button-options editBottom"
                    onClick={() => handleOpenModal({ product })}
                >
                    <FaEdit />
                </button>
                <button
                    className="button-options deleteBottom"
                    onClick={handleDelete}  // Llamamos a handleDelete al hacer clic
                >
                    <FaTrash />
                </button>
            </td>
        </tr>
    );
};
