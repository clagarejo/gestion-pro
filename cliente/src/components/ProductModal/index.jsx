import React from 'react';
import { useProductForm } from '@/hooks/useProductForm';
import { ProductForm } from './ProductForm';
import { useProductStore } from '@/store/useProductStore';
import './styles.scss';

export const ProductModal = ({ product, onClose }) => {
    const { formData, handleChange, validateForm } = useProductForm(product);
    const { addProduct, updateProduct } = useProductStore();

    const handleSubmit = async (e) => {
        e.preventDefault();

        if (!validateForm()) return;

        try {
            if (product) {
                updateProduct(product.product._id, formData);
            } else {
                addProduct(formData);
            }

            onClose();
        } catch (error) {
            console.error('Error al crear o actualizar el producto:', error);
            Swal.fire('Error', 'Hubo un problema al procesar tu solicitud.', 'error');
        }
    };

    return (
        <div className="modal active">
            <div className="modal-content">
                <div className="modal-header">
                    <h3>{product ? 'Editar producto' : 'Crear producto'}</h3>
                    <button className="close-btn" onClick={onClose}>X</button>
                </div>
                <ProductForm
                    formData={formData}
                    handleChange={handleChange}
                    handleSubmit={handleSubmit}
                />
            </div>
        </div>
    );
};
