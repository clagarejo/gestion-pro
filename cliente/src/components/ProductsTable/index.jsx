import { useState } from 'react';
import { ProductModal } from '../ProductModal';
import { Product } from '../Product';
import './styles.scss';

export const ProductsTable = () => {
    const [isModalOpen, setIsModalOpen] = useState(false);
    const [productToEdit, setProductToEdit] = useState(null);
    const [currentPage, setCurrentPage] = useState(1);
    const itemsPerPage = 5;

    const handleOpenModal = (product = null) => {
        setProductToEdit(product);
        setIsModalOpen(true);
    };

    const handleCloseModal = () => {
        setIsModalOpen(false);
        setProductToEdit(null);
    };

    const products = [
        { name: 'Producto 1', category: 'Electrónica', price: 100, stock: 20 },
        { name: 'Producto 2', category: 'Ropa', price: 25, stock: 50 },
        { name: 'Producto 3', category: 'Alimentos', price: 15, stock: 80 },
        { name: 'Producto 4', category: 'Hogar', price: 75, stock: 10 },
        { name: 'Producto 5', category: 'Deportes', price: 45, stock: 5 },
        { name: 'Producto 6', category: 'Oficina', price: 30, stock: 25 },
        { name: 'Producto 7', category: 'Mascotas', price: 20, stock: 40 },
        { name: 'Producto 8', category: 'Juguetes', price: 50, stock: 15 },
    ];

    const totalPages = Math.ceil(products.length / itemsPerPage);
    const paginatedProducts = products.slice(
        (currentPage - 1) * itemsPerPage,
        currentPage * itemsPerPage
    );

    const handlePageChange = (page) => {
        setCurrentPage(page);
    };

    return (
        <div className="container">
            <div className="container_bottom">
                <input
                    className="search-products"
                    type="text"
                    placeholder="Buscar productos"
                />
                <button onClick={() => handleOpenModal()}>Crear producto</button>
            </div>

            <table className="table">
                <thead>
                    <tr>
                        <th>Nombre del Producto</th>
                        <th>Categoría</th>
                        <th>Precio</th>
                        <th style={{ textAlign: 'center' }}>Cantidad en Stock</th>
                        <th style={{ textAlign: 'center' }}>Opciones</th>
                    </tr>
                </thead>
                <tbody>
                    {paginatedProducts.map((product, index) => (
                        <Product key={index} {...product} handleOpenModal={handleOpenModal} />
                    ))}
                </tbody>
            </table>

            <div className="pagination">
                {Array.from({ length: totalPages }, (_, index) => (
                    <button
                        key={index}
                        className={`pagination-button ${currentPage === index + 1 ? 'active' : ''}`}
                        onClick={() => handlePageChange(index + 1)}
                    >
                        {index + 1}
                    </button>
                ))}
            </div>

            {isModalOpen && (
                <ProductModal
                    product={productToEdit}
                    onClose={handleCloseModal}
                />
            )}
        </div>
    );
};
