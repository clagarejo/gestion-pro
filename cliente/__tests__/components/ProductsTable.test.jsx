import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { ProductsTable } from '@/components/ProductsTable';
import { useProducts } from '@/hooks/useProducts';
// import '@testing-library/jest-dom/extend-expect';

// Mocking el hook useProducts
jest.mock('@/hooks/useProducts', () => ({
    useProducts: jest.fn(),
}));

describe('ProductsTable', () => {
    const mockSearchProducts = jest.fn();
    const mockHandleFileUpload = jest.fn();

    beforeEach(() => {
        useProducts.mockReturnValue({
            products: [
                { _id: '1', name: 'Producto 1', category: 'Categoría 1', price: '10', stock: '100' },
                { _id: '2', name: 'Producto 2', category: 'Categoría 2', price: '20', stock: '50' },
            ],
            selected: [],
            loading: false,
            searchProducts: mockSearchProducts,
            handleFileUpload: mockHandleFileUpload,
        });
    });

    test('debe manejar el cambio en el input de búsqueda', () => {
        render(<ProductsTable />);

        const searchInput = screen.getByPlaceholderText('Buscar productos');
        fireEvent.change(searchInput, { target: { value: 'Producto 1' } });

        expect(mockSearchProducts).toHaveBeenCalledWith('Producto 1');
    });


    test('debe coincidir con el snapshot', () => {
        const { asFragment } = render(<ProductsTable />);
        expect(asFragment()).toMatchSnapshot();
    });
});
