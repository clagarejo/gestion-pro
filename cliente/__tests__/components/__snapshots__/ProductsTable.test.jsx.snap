// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsTable debe coincidir con el snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <header
      class="container_header"
    >
      <h1>
        Gestión Pro
      </h1>
      <button
        class="theme-toggle"
      >
        ☀️ 
      </button>
    </header>
    <div
      class="container_bottom"
    >
      <input
        class="search-products"
        placeholder="Buscar productos"
        type="text"
        value=""
      />
      <div
        class="action-buttons"
      >
        <button
          class="massive-upload"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M537.6 226.6c4.1-10.7 6.4-22.4 6.4-34.6 0-53-43-96-96-96-19.7 0-38.1 6-53.3 16.2C367 64.2 315.3 32 256 32c-88.4 0-160 71.6-160 160 0 2.7.1 5.4.2 8.1C40.2 219.8 0 273.2 0 336c0 79.5 64.5 144 144 144h368c70.7 0 128-57.3 128-128 0-61.9-44-113.6-102.4-125.4zM393.4 288H328v112c0 8.8-7.2 16-16 16h-48c-8.8 0-16-7.2-16-16V288h-65.4c-14.3 0-21.4-17.2-11.3-27.3l105.4-105.4c6.2-6.2 16.4-6.2 22.6 0l105.4 105.4c10.1 10.1 2.9 27.3-11.3 27.3z"
            />
          </svg>
           Cargar productos
        </button>
        <input
          style="display: none;"
          type="file"
        />
        <button
          class="add_product"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="margin-right: 10px;"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
            />
          </svg>
           Crear producto
        </button>
      </div>
    </div>
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            Nombre del Producto
          </th>
          <th>
            Categoría
          </th>
          <th>
            Precio
          </th>
          <th>
            Cantidad en Stock
          </th>
          <th>
            Opciones
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Producto 1
          </td>
          <td>
            Categoría 1
          </td>
          <td>
            $ 10(COP)
          </td>
          <td>
            100
          </td>
          <td
            class="tr-options"
          >
            <button
              class="button-options editBottom"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
                />
              </svg>
            </button>
            <button
              class="button-options deleteBottom"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                />
              </svg>
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Producto 2
          </td>
          <td>
            Categoría 2
          </td>
          <td>
            $ 20(COP)
          </td>
          <td>
            50
          </td>
          <td
            class="tr-options"
          >
            <button
              class="button-options editBottom"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
                />
              </svg>
            </button>
            <button
              class="button-options deleteBottom"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                />
              </svg>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="pagination"
    >
      <button
        class="pagination-button active"
      >
        1
      </button>
    </div>
  </div>
</DocumentFragment>
`;
